### PR TITLE
applications: nrf5340_audio: Move controller

### PR DESCRIPTION
**Pull request by koffes**

### CODE REVIEW

**Review by greg-fer — CHANGES_REQUESTED** [review]
This is a major doc PR, and unfortunately it came unexpectedly in the week of the RC-1 for NCS 2.3.0. Hence I can offer only some general comments for the docs for now, as I am busy with other doc tasks.

Comments:

- ib/bin/bt_ll_pcft:
  
  - README.rst from l needs to be removed. Libraries are only to be documented under doc/nrf/libraries.

- doc/nrf/libraries/bin/bt:

  - CHANGELOG.rst should not link to release notes for known issues, but to known_issues.rst. Also, limitations should be listed in index.rst.
  - index.rst must match the template for libraries: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/templates/library_template_README.html -- the sections without the asterisk are mandatory, so they must be included + the section about Limitations (see comment above)
  - the ".. |net_core_hex_note| replace::" lines must go back to the application's README.rst. Moving them out of that file will break all the replaces in the application doc.

**Comment by koffes** [comment]
@greg-fer I have based the docs on the template and moved the README. Ping me at your convenience.

File: applications/nrf5340_audio/CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,7 @@
 cmake_minimum_required(VERSION 3.20.0)
 
 file(TO_CMAKE_PATH $ENV{ZEPHYR_BASE} CMAKE_STYLE_ZEPHYR)
+file(TO_CMAKE_PATH ${CMAKE_CURRENT_LIST_DIR}/../../lib/bin/bt_ll_pcft/bin CMAKE_STYLE_LL_PCFT_BIN)
```
Comment (alexsven) [inline applications/nrf5340_audio/CMakeLists.txt:10]:
```suggestion
file(TO_CMAKE_PATH ${ZEPHYR_NRF_MODULE_DIR}/lib/bin/bt_ll_pcft/bin CMAKE_STYLE_LL_PCFT_BIN)
```

File: doc/nrf/libraries/bin/bt_ll_pcft/index.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
+.. _lib_bt_ll_pcft_readme:
+
+LE Audio Controller for nRF5340 (Net core controller)
+#####################################################
+
+.. contents::
+   :local:
+   :depth: 2
+
+This link layer controller for the nRF5340 NET core, is used for LE Audio applications.
```
Comment (alexsven) [inline doc/nrf/libraries/bin/bt_ll_pcft/index.rst:10]:
```suggestion
This link layer controller, for the nRF5340 NET core, is used for LE Audio applications.
```

File: doc/nrf/libraries/bin/bt_ll_pcft/index.rst [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+This controller is given as a binary (hex file) and is provided by Packetcraft.
+
+Overview
+********
+Communication between the APP and NET core is handled by the HCIRPMsg library.
```
Comment (alexsven) [inline doc/nrf/libraries/bin/bt_ll_pcft/index.rst:15]:
```suggestion
Communication between the APP and NET core is handled by the HCI_RPMsg library.
```

File: doc/nrf/libraries/bin/bt_ll_pcft/index.rst [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+Files
+*****
+.. |net_core_hex_note| replace:: The network core for both gateway and headsets are programmed with the precompiled Bluetooth Low Energy Controller binary file :file:`ble5-ctr-rpmsg_<X>.hex`, where *<X>* corresponds to the controller version, for example :file:`ble5-ctr-rpmsg_3216.hex`.
+   This file includes the LE Audio Controller for nRF53 and is provided in the :file:`/bin` directory.
+   If :ref:`DFU is enabled <nrf53_audio_app_configuration_configure_fota>`, the binary file will be generated in the :file:`build/zephyr/` directory and will be called :file:`net_core_app_signed.hex`.
+
+There are two other files in the :file:`/bin` directory:
+:file:`ble5-ctr-rpmsg_shifted_<X>.hex` which has been shifted to make space for mcuboot.
+:file:`ble5-ctr-rpmsg_shifted_min_<X>.hex`which is the same as above, but made for minimum mcuboot.
```
Comment (alexsven) [inline doc/nrf/libraries/bin/bt_ll_pcft/index.rst:41]:
```suggestion
:file:`ble5-ctr-rpmsg_shifted_min_<X>.hex` which is the same as above, but made for minimum mcuboot.
```

**Review by alexsven — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/CMakeLists.txt [hunk 5]
```diff
@@ -96,7 +97,7 @@ if ((CONFIG_AUDIO_DFU EQUAL 1) OR (CONFIG_AUDIO_DFU EQUAL 2))
```
Comment (pirun) [inline applications/nrf5340_audio/CMakeLists.txt:97]: Should move to here.
file(TO_CMAKE_PATH ${ZEPHYR_NRF_MODULE_DIR}/lib/bin/bt_ll_pcft/bin CMAKE_STYLE_LL_PCFT_BIN)

File: lib/bin/bt_ll_pcft/include/ble_hci_vsc.h [hunk 6]
```diff
@@ -4,6 +4,13 @@
...
 
+/**
+ * @file ble_hci_vsc.h
+ *
+ * @defgroup ble_hci_vsc_api LE Audio Controller for nRF5340 API
+ * @{
+ */
+
 #ifndef _BLE_HCI_VSC_H_
 #define _BLE_HCI_VSC_H_
```
Comment (greg-fer) [inline lib/bin/bt_ll_pcft/include/ble_hci_vsc.h:16]: Before I review this header, I see that it is incomplete, as the defines, structs and enums in lines 19-111 are not documented. 

@koffes , can you add API documentation for these elements?

File: applications/nrf5340_audio/README.rst [hunk 7]
```diff
@@ -56,7 +56,7 @@ Regardless of the configuration, the application handles the audio data in the f
    #. Audio data is encoded by the software codec.
    #. Encoded audio data is sent to the Bluetooth LE Host.
 
-#. The host sends the encoded audio data to the LE Audio Controller Subsystem for nRF53 on the network core.
+#. The host sends the encoded audio data to the LE Audio Controller Subsystem for nRF53 on the network core. <TODO: Add link>
```
Comment (greg-fer) [inline applications/nrf5340_audio/README.rst:59]: Note to self: @greg-fer , review in the next round.

File: doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+Notable changes to this controller are documented in this file.
+
+Controller v3322
+****************
+
+Changes
+=======
+* Changes to accommodate BIS + ACL combinations.
+* Improvements to support creating CIS connections in any order.
+* Basic support for interleaved broadcasts
```
Comment (greg-fer) [inline doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst:19]: Note to self: @greg-fer , review in the next round.

File: lib/bin/bt_ll_pcft/Kconfig [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+config BT_LL_PCFT
```
Comment (greg-fer) [inline lib/bin/bt_ll_pcft/Kconfig:7]: Note to self: @greg-fer , add this to the readme.

**Review by greg-fer — CHANGES_REQUESTED** [review]

**Comment by koffes** [comment]
@asbjornsabo, no need to review the entire PR. I just want to check that you agree that the added PCFT controller license is correct.

File: doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+Changelog
+#########
+
+.. contents::
+   :local:
+   :depth: 2
+
+Notable changes to this controller are documented in this file.
+
+Controller version 3322
```
Comment (asbjornsabo) [inline doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst:12]: I don't understand.
The 3330 controller is moved to a new place.  And a changelog is added for the 3322 controller.

File: doc/nrf/libraries/bin/bt_ll_pcft/LICENSE.txt [hunk 11]
```diff
@@ -0,0 +1,32 @@
+Copyright (c) 2021, PACKETCRAFT, INC.
```
Comment (asbjornsabo) [inline doc/nrf/libraries/bin/bt_ll_pcft/LICENSE.txt:1]: This is the license text that the License and Service Agreement mandates, so this is correct.

**Review by asbjornsabo — COMMENTED** [review]
The controller license document is the correct one.

There seems to be some issue with controller versions in the cghanbgelog.

**Comment by greg-fer** [comment]
Looks like all my edits were overwritten :(

**Review by greg-fer — APPROVED** [review]

File: doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+* Basic support for interleaved broadcasts.
+
+Known issues
+************
+
+See the :ref:`nRF5340 Audio application known issues <known_issues_nrf5340audio>` for the list of known issues for the controller.
+
+Limitations
+***********
+
```
Comment (koffes) [inline doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst:31]: Add link to functionality table.

File: lib/bin/bt_ll_pcft/include/ble_hci_vsc.h [hunk 13]
```diff
@@ -90,75 +108,77 @@ enum ble_hci_vs_tx_power {
...
 
 /**
- * @brief Set TX power for specific connection
- * @param conn_handle Specific connection handle for TX power setting
- * @param tx_power TX power setting for the specific connection handle
- *                 Please check ble_hci_vs_tx_power for possible settings
+ * @brief Set TX power for specific connection.
+ * @param conn_handle Specific connection handle for the TX power setting.
+ * @param tx_power TX power setting for the specific connection handle.
+ *                 Check ble_hci_vs_tx_power for possible settings
```
Comment (erikrobstad) [inline lib/bin/bt_ll_pcft/include/ble_hci_vsc.h:181]:
```suggestion
 *                 Check ble_hci_vs_tx_power for possible settings.
```

**Review by erikrobstad — CHANGES_REQUESTED** [review]

**Comment by pirun** [comment]
@koffes My previous change is overwritten. Please squash the variable correction commit.

**Review by alexsven — APPROVED** [review]

File: lib/bin/bt_ll_pcft/include/ble_hci_vsc.h (hunk 13 above)
Comment (koffes) [reply lib/bin/bt_ll_pcft/include/ble_hci_vsc.h]: Done

File: lib/bin/bt_ll_pcft/include/ble_hci_vsc.h [hunk 14]
```diff
@@ -33,46 +41,56 @@ struct ble_hci_vs_cp_nrf21540_pins {
...
 } __packed;
 
+/** @brief Host-Controller Interface vendor-specific structure for setting custom flags. */
 struct ble_hci_vs_cp_set_op_flag {
 	uint32_t flag_bit;
 	uint8_t setting;
 } __packed;
 
+/** @brief Host-Controller Interface vendor-specific Bluetooth address. */
+
```
[the remark this answers is not in the document]
Comment (koffes) [reply lib/bin/bt_ll_pcft/include/ble_hci_vsc.h]: Done

File: lib/bin/CMakeLists.txt [hunk 15]
```diff
@@ -4,4 +4,5 @@
 # SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
 #
 
-add_subdirectory_ifdef(CONFIG_LWM2M_CARRIER lwm2m_carrier)
+add_subdirectory_ifdef(CONFIG_LWM2M_CARRIER         lwm2m_carrier)
```
[the remark this answers is not in the document]
Comment (koffes) [reply lib/bin/CMakeLists.txt]: No. Changed.

File: doc/nrf/libraries/bin/bt_ll_pcft/index.rst [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+Files
+*****
+
+The network core for both gateway and headsets of the :ref:`nrf53_audio_app` is programmed with the precompiled Bluetooth Low Energy Controller binary file :file:`ble5-ctr-rpmsg_<XYZ>.hex`, where *<XYZ>* corresponds to the controller version, for example :file:`ble5-ctr-rpmsg_3216.hex`.
+This file includes the LE Audio Controller Subsystem for nRF53.
+If :ref:`DFU is enabled <nrf53_audio_app_configuration_configure_fota>`, the subsystem's binary file will be generated in the :file:`build/zephyr/` directory and will be called :file:`net_core_app_signed.hex`.
+
+There are two other files in the :file:`/bin` directory:
+
+* :file:`ble5-ctr-rpmsg_shifted_<XYZ>.hex` which has been shifted to make space for mcuboot.
```
[the remark this answers is not in the document]
Comment (koffes) [reply doc/nrf/libraries/bin/bt_ll_pcft/index.rst]: Done

File: applications/nrf5340_audio/tools/buildprog/buildprog.py [hunk 17]
```diff
@@ -179,24 +180,21 @@ def __populate_hex_paths(dev, options):
...
             dest_folder = TARGET_CORE_NET_FOLDER
 
-            if len(hex_files_found) == 0:
-                raise Exception(
-                    f"Found no net core hex file in folder: {dest_folder}")
-            elif len(hex_files_found) > 1:
+            if hex_files_found != 1:
                 raise Exception(
-                    f"Found more than one hex file in folder: {dest_folder}")
+                    f"Did not find a single matching NET hex file in folder: {dest_folder}")
```
[the remark this answers is not in the document]
Comment (koffes) [reply applications/nrf5340_audio/tools/buildprog/buildprog.py]: Done

**Review by Frodevan — DISMISSED** [review]
Test system needs to be adjusted before this is merged

File: applications/nrf5340_audio/CMakeLists.txt [hunk 18]
```diff
@@ -7,6 +7,7 @@
 cmake_minimum_required(VERSION 3.20.0)
 
 file(TO_CMAKE_PATH $ENV{ZEPHYR_BASE} CMAKE_STYLE_ZEPHYR)
+file(TO_CMAKE_PATH ${ZEPHYR_NRF_MODULE_DIR}/lib/bin/bt_ll_pcft/bin CMAKE_STYLE_LL_PCFT_BIN)
```
[the remark this answers is not in the document]
Comment (koffes) [reply applications/nrf5340_audio/CMakeLists.txt]: Did you look at the latest version of the code? This was changed a few days ago.

File: doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst (hunk 12 above)
Comment (koffes) [reply doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst]: @greg-fer please have a look and se if you find this ok.

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+Known issues
+************
+
+See the :ref:`nRF5340 Audio application known issues <known_issues_nrf5340audio>` for the list of known issues for the controller.
+
+Limitations
+***********
+
+The controller is closely related to the nRF5340 Audio application. Please see :ref:`_software_maturity_application_nrf5340audio_table`.
```
Comment (greg-fer) [inline doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst:32]:
```suggestion
The controller is closely related to the nRF5340 Audio application.
See :ref:`software_maturity_application_nrf5340audio_table` for the list of supported features and limitations.
```

**Review by greg-fer — CHANGES_REQUESTED** [review]
One comment.

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+Changelog
+#########
+
+.. contents::
+   :local:
+   :depth: 2
+
+Notable changes to this controller are documented in this file.
+
+Controller v3322
```
Comment (erikrobstad) [inline doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst:12]: Like @asbjornsabo previosly mentioned, the controller moved is not 3322, but 3330.
Is it correct to list v3322 here?

**Review by erikrobstad — CHANGES_REQUESTED** [review]

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst (hunk 20 above)
Comment (greg-fer) [reply doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst:12]: Or rather: shouldn't we also list v3330 here, in addition to v3322? I would keep v3322, if it was used at some point and we have changelog changes documented for it.

File: doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst (hunk 10 above)
Comment (koffes) [reply doc/nrf/libraries/bin/bt_ll_pcft/CHANGELOG.rst]: I have added changelog for both now.

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst (hunk 19 above)
Comment (koffes) [reply doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst]: Done

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst (hunk 20 above)
Comment (koffes) [reply doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst:34]: Done

**Review by erikrobstad — APPROVED** [review]

**Review by greg-fer — APPROVED** [review]

**Review by greg-fer — APPROVED** [review]

**Pull request merged**